mw_transform_jq.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/Sirupsen/logrus"
+	"github.com/mitchellh/mapstructure"
+	"io/ioutil"
+	"net/http"
+)
+
+type TransformJQMiddleware struct {
+	*BaseMiddleware
+}
+
+type JQTransformOptions struct {
+	OutputHeaders map[string]string `mapstructure:"output_headers"`
+	OutputVars    map[string]string `mapstructure:"output_vars"`
+}
+
+func (t *TransformJQMiddleware) GetName() string {
+	return "TransformJQMiddleware"
+}
+
+func (t *TransformJQMiddleware) IsEnabledForSpec() bool {
+	for _, version := range t.Spec.VersionData.Versions {
+		if len(version.ExtendedPaths.TransformJQ) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (t *TransformJQMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	_, versionPaths, _, _ := t.Spec.GetVersionData(r)
+	found, meta := t.Spec.CheckSpecMatchesStatus(r, versionPaths, TransformedJQ)
+	if !found {
+		return nil, 200
+	}
+	err := transformJQBody(r, meta.(*TransformJQSpec), t.Spec.EnableContextVars)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"prefix":      "inbound-transform-jq",
+			"server_name": t.Spec.Proxy.TargetURL,
+			"api_id":      t.Spec.APIID,
+			"path":        r.URL.Path,
+		}).Error(err)
+		return err, 415
+	}
+	return nil, 200
+}
+
+func transformJQBody(r *http.Request, t *TransformJQSpec, contextVars bool) error {
+	// Read the body:
+	defer r.Body.Close()
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return err
+	}
+
+	var contextJson []byte
+	if contextVars {
+		contextJson, _ = json.Marshal(ctxGetData(r))
+	} else {
+		contextJson = []byte("{}")
+	}
+
+	bodyBuffer := bytes.NewBufferString("[")
+	bodyBuffer.Write(contextJson)
+	bodyBuffer.WriteString(",")
+	bodyBuffer.Write(body)
+	bodyBuffer.WriteString("]")
+
+	err = t.JQFilter.HandleJson(bodyBuffer.String())
+	if err != nil {
+		return errors.New("Input is not a valid JSON")
+	}
+
+	// First element is the transformed body
+	if t.JQFilter.Next() {
```

Comment:
I'm not sure why we have 2 steps here (t.JQFilter.Next()). Can you clarify it?

Reply:
The first element is the actual transformed body transformed by JQ.

If the JQ filter provides a second element (another JSON object) with specific "output_headers" and/or "output_vars" object, it will be given back to tyk in order to update he headers of the tyk query/reponse, or feed the context vars.

For example, let's take the following jq filter:

    {mynewfield: .data}

it will return a JSON object with only the data field.

Now, if the filter is:

    {mynewfield: .data}, {output_headers: {x_name: .name}, output_vars: {age: .userAge}}

it will add the `x_name` header to the http query or response, and add the `age` variable to the context vars, which will be available with the `$tyk_context.jq_output_var_age` in url rewrites for example.

Thinking of it, I now think that the output_headers is useless, and that the output_vars mechanism can be used to add specific HTTP headers (if the modify_headers middleware is context vars aware).

Reply:
Missed your message! 

Thank you for clarification, I was confused because our transformation middleware supports only response or request body modification. And your one also supports headers and context variables.

I'm not sure if fully understand use case of setting context variables or headers based on the body. Is it something you actually use by your own?

Reply:
Yes, it's a use case we can have. I understand it's very specific to our usecase.
For example, we can imagine having a "deviceID" field inside the JSON body, that we want to put as a X-DEVICE-ID http header for the upstream server.

Reply:
Additional queston, in your examples you do: `. as [$reqContext, $resHeaders, $body] `, which is a bit cryptic. Can we use proxy object instead, smth like .tyk.body, .tyk.context and etc?

Reply:
This is a way to give metadata (like the headers or the request context) other than the body itself, to the jq filter.

So instead of just giving the actual body to the jq filter, Tyk gives the body, inside a 3 elements array, along with the requestContext and the response headers.

That is why, the jq filter have to begin with `. as [$reqContext, $resHeaders, $body]` in order to assign each element of the array to a variable that we can further use in the jq filter.

I agree it's a trick, and I found no other way to do it at the moment.

Reply:
Not sure if you got my idea. `.` in jq treated as a `this` or input data. What if instead of passing array to it, we will pass `{ "body": .., "context": ... , "headers": ... }` object, so our jq filter will look like:
```
.body + {\"TRANSFORMED-REQUEST-BY-JQ\": true, path: .context.path, user_agent: .context.headers_User_Agent}```

Reply:
Also, just to clarify, in jq syntax "+" means merge objects, right?

Reply:
Yes, "+" means merge.

Regarding your idea, yes, we can do that as well. But at the end, it's the same idea, except that instead of an array (an maybe a cryptic syntax to do variable assignment), it's a object.

I'll update my code, to use an object instead of an array, it will make things easier to understand.

Good idea !